File: modules/expo-mlkit-face-detector/android/src/main/java/expo/modules/mlkitfacedetector/ExpoMlkitFaceDetectorModule.kt
package expo.modules.mlkitfacedetector

import android.content.Context
import android.graphics.PointF
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class ExpoMlkitFaceDetectorModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React Application Context is null")

  override fun definition() = ModuleDefinition {
    Name("ExpoMlkitFaceDetector")

    AsyncFunction("detectFacesAsync") { imageUriString: String, optionsMap: Map<String, Any>?, promise: Promise ->
      try {
        val uri = if (imageUriString.startsWith("file://") || imageUriString.startsWith("content://")) {
          Uri.parse(imageUriString)
        } else {
          Uri.fromFile(File(imageUriString))
        }

        val inputImage = InputImage.fromFilePath(context, uri)

        // Parse options
        val modeStr = optionsMap?.get("mode") as? String ?: "accurate"
        val detectLandmarks = optionsMap?.get("detectLandmarks") as? Boolean ?: true
        val classifyModeStr = optionsMap?.get("classifyMode") as? String ?: "all"
        val minFaceSize = (optionsMap?.get("minFaceSize") as? Number)?.toFloat() ?: 0.15f
        val tracking = optionsMap?.get("tracking") as? Boolean ?: false

        val builder = FaceDetectorOptions.Builder()
          .setMinFaceSize(minFaceSize)

        if (modeStr == "fast") {
          builder.setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
        } else {
          builder.setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
        }

        if (detectLandmarks) {
          builder.setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
        } else {
          builder.setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
        }

        if (classifyModeStr == "all") {
          builder.setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
        } else {
          builder.setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
        }

        if (tracking) {
          builder.enableTracking()
        }

        val detector = FaceDetection.getClient(builder.build())

        detector.process(inputImage)
          .addOnSuccessListener { faces ->
            val resultFaces = ArrayList<Map<String, Any?>>()

            for (face in faces) {
              val faceMap = HashMap<String, Any?>()
              val bounds = face.boundingBox

              val boundsMap = mapOf(
                "x" to bounds.left,
                "y" to bounds.top,
                "width" to bounds.width(),
                "height" to bounds.height()
              )
              faceMap["bounds"] = boundsMap
              faceMap["rollAngle"] = face.headEulerAngleZ.toDouble() // roll
              faceMap["yawAngle"] = face.headEulerAngleY.toDouble()  // yaw
              faceMap["pitchAngle"] = face.headEulerAngleX.toDouble() // pitch
              faceMap["smilingProbability"] = face.smilingProbability?.toDouble()
              faceMap["leftEyeOpenProbability"] = face.leftEyeOpenProbability?.toDouble()
              faceMap["rightEyeOpenProbability"] = face.rightEyeOpenProbability?.toDouble()
              faceMap["trackingId"] = face.trackingId

              // Landmarks
              val landmarksMap = HashMap<String, Any?>()
              face.getLandmark(FaceLandmark.LEFT_EYE)?.let {
                landmarksMap["leftEye"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.RIGHT_EYE)?.let {
                landmarksMap["rightEye"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.LEFT_EAR)?.let {
                landmarksMap["leftEar"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.RIGHT_EAR)?.let {
                landmarksMap["rightEar"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.LEFT_CHEEK)?.let {
                landmarksMap["leftCheek"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.RIGHT_CHEEK)?.let {
                landmarksMap["rightCheek"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.NOSE_BASE)?.let {
                landmarksMap["noseBase"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.MOUTH_LEFT)?.let {
                landmarksMap["mouthLeft"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.MOUTH_RIGHT)?.let {
                landmarksMap["mouthRight"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              face.getLandmark(FaceLandmark.MOUTH_BOTTOM)?.let {
                landmarksMap["mouthBottom"] = mapOf("x" to it.position.x, "y" to it.position.y)
              }
              faceMap["landmarks"] = landmarksMap

              resultFaces.add(faceMap)
            }

            val resultMap = mapOf(
              "hasFace" to (resultFaces.isNotEmpty()),
              "faceCount" to resultFaces.size,
              "faces" to resultFaces,
              "imageWidth" to inputImage.width,
              "imageHeight" to inputImage.height
            )

            promise.resolve(resultMap)
          }
          .addOnFailureListener { exception ->
            promise.reject("MLKIT_FACE_DETECTION_ERROR", exception.message ?: "Failed to detect faces with ML Kit", exception)
          }
          .addOnCompleteListener {
            detector.close()
          }
      } catch (e: Exception) {
        promise.reject("MLKIT_FACE_DETECTION_EXCEPTION", e.message ?: "Unexpected error during face detection", e)
      }
    }
  }
}

File: modules/expo-mlkit-face-detector/ios/ExpoMlkitFaceDetectorModule.swift
import ExpoModulesCore
import MLKitFaceDetection
import MLKitVision
import UIKit

public class ExpoMlkitFaceDetectorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoMlkitFaceDetector")

    AsyncFunction("detectFacesAsync") { (imageUriString: String, optionsDict: [String: Any]?, promise: Promise) in
      guard let url = URL(string: imageUriString),
            let data = try? Data(contentsOf: url),
            let uiImage = UIImage(data: data) else {
        promise.reject("INVALID_IMAGE_URI", "Could not load image from URI: \(imageUriString)")
        return
      }

      let visionImage = VisionImage(image: uiImage)
      visionImage.orientation = uiImage.imageOrientation

      let options = FaceDetectorOptions()
      let modeStr = (optionsDict?["mode"] as? String) ?? "accurate"
      options.performanceMode = (modeStr == "fast") ? .fast : .accurate
      
      let detectLandmarks = (optionsDict?["detectLandmarks"] as? Bool) ?? true
      options.landmarkMode = detectLandmarks ? .all : .none

      let classifyModeStr = (optionsDict?["classifyMode"] as? String) ?? "all"
      options.classificationMode = (classifyModeStr == "all") ? .all : .none

      if let minFaceSize = optionsDict?["minFaceSize"] as? NSNumber {
        options.minFaceSize = CGFloat(minFaceSize.doubleValue)
      }

      if let tracking = optionsDict?["tracking"] as? Bool, tracking {
        options.isTrackingEnabled = true
      }

      let faceDetector = FaceDetector.faceDetector(options: options)

      faceDetector.process(visionImage) { faces, error in
        if let error = error {
          promise.reject("MLKIT_FACE_DETECTION_ERROR", error.localizedDescription)
          return
        }

        var resultFaces: [[String: Any?]] = []

        for face in (faces ?? []) {
          var faceMap: [String: Any?] = [:]
          let frame = face.frame

          faceMap["bounds"] = [
            "x": frame.origin.x,
            "y": frame.origin.y,
            "width": frame.size.width,
            "height": frame.size.height
          ]
          faceMap["rollAngle"] = face.headEulerAngleZ
          faceMap["yawAngle"] = face.headEulerAngleY
          faceMap["pitchAngle"] = face.headEulerAngleX
          faceMap["smilingProbability"] = face.hasSmilingProbability ? face.smilingProbability : nil
          faceMap["leftEyeOpenProbability"] = face.hasLeftEyeOpenProbability ? face.leftEyeOpenProbability : nil
          faceMap["rightEyeOpenProbability"] = face.hasRightEyeOpenProbability ? face.rightEyeOpenProbability : nil
          faceMap["trackingId"] = face.hasTrackingID ? face.trackingID : nil

          var landmarksMap: [String: Any] = [:]
          if let leftEye = face.landmark(ofType: .leftEye) {
            landmarksMap["leftEye"] = ["x": leftEye.position.x, "y": leftEye.position.y]
          }
          if let rightEye = face.landmark(ofType: .rightEye) {
            landmarksMap["rightEye"] = ["x": rightEye.position.x, "y": rightEye.position.y]
          }
          if let leftEar = face.landmark(ofType: .leftEar) {
            landmarksMap["leftEar"] = ["x": leftEar.position.x, "y": leftEar.position.y]
          }
          if let rightEar = face.landmark(ofType: .rightEar) {
            landmarksMap["rightEar"] = ["x": rightEar.position.x, "y": rightEar.position.y]
          }
          if let noseBase = face.landmark(ofType: .noseBase) {
            landmarksMap["noseBase"] = ["x": noseBase.position.x, "y": noseBase.position.y]
          }
          if let mouthLeft = face.landmark(ofType: .mouthLeft) {
            landmarksMap["mouthLeft"] = ["x": mouthLeft.position.x, "y": mouthLeft.position.y]
          }
          if let mouthRight = face.landmark(ofType: .mouthRight) {
            landmarksMap["mouthRight"] = ["x": mouthRight.position.x, "y": mouthRight.position.y]
          }
          if let mouthBottom = face.landmark(ofType: .mouthBottom) {
            landmarksMap["mouthBottom"] = ["x": mouthBottom.position.x, "y": mouthBottom.position.y]
          }
          faceMap["landmarks"] = landmarksMap

          resultFaces.append(faceMap)
        }

        let result: [String: Any] = [
          "hasFace": !resultFaces.isEmpty,
          "faceCount": resultFaces.count,
          "faces": resultFaces,
          "imageWidth": uiImage.size.width,
          "imageHeight": uiImage.size.height
        ]

        promise.resolve(result)
      }
    }
  }
}
